### ISSUE TITLE
Issues updating from 0.1.1 to 0.2.0

### ISSUE THREAD

**Issue by dnlldl** [tensorchord/pgvecto.rs#379]
Immich user here, trying to update pgvecto.rs from 0.1.1 to 0.2.0, using pg14. Here is what I did on my system:

```
postgres@vm-dev:~> cd /tmp
postgres@vm-dev:/tmp> wget https://github.com/tensorchord/pgvecto.rs/archive/refs/tags/v0.2.0.zip
postgres@vm-dev:/tmp> unzip ./v0.2.0.zip
postgres@vm-dev:/tmp> cd pgvecto.rs-0.2.0/
postgres@vm-dev:/tmp/pgvecto.rs-0.2.0> cargo pgrx install -c /usr/lib/postgresql14/bin/pg_config --sudo --release
```
However, I noticed in `/var/lib/pgsql/pgvecto.rs-0.2.0/Cargo.toml` that it says `version = "0.0.0"`. I changed it manually to `0.2.0` and tried again. Which seemed to work... however, now, when I'm trying to update it in my database:

```
immich=# ALTER EXTENSION vectors UPDATE TO '0.2.0';
ERROR:  extension "vectors" has no update path from version "0.1.1" to version "0.2.0"
```

```
immich=# select * from pg_available_extensions where name = 'vectors';
  name   | default_version | installed_version |                                           comment                                  
---------+-----------------+-------------------+----------------------------------------------------------------------------------------------
 vectors | 0.2.0           | 0.1.1             | vectors: Vector database plugin for Postgres, written in Rust, specifically designed for LLM
(1 row)
```

So I tried removing it...

```
immich=# drop extension vectors;
ERROR:  cannot drop extension vectors because other objects depend on it
DETAIL:  column embedding of table asset_faces depends on type vector
column embedding of table smart_search depends on type vector
HINT:  Use DROP ... CASCADE to drop the dependent objects too.

immich=# drop extension vectors cascade;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to column embedding of table asset_faces
drop cascades to column embedding of table smart_search
ERROR:  could not find function "vectors_amhandler_wrapper" in file "/usr/lib/postgresql14/lib64/vectors.so"
```
And this is where I'm currently stucked. I'm writing a guide in my own wiki to be able to update next time, this is far from easy when using a dedicated server rather than dockers.

**Comment by VoVAllen** [comment]
Sorry for the inconvenience. We didn't notice before that immich also supports bare metal deployment. 

To install pgvecto.rs, you don't need to build from source. We have the deb release package at https://github.com/tensorchord/pgvecto.rs/releases/tag/v0.2.0.

- Have you tried the command at https://docs.pgvecto.rs/admin/upgrading.html#upgrade-from-0-1-x with 
```SQL
UPDATE pg_catalog.pg_extension SET extversion = '0.1.x' where extname = 'vectors';
UPDATE pg_catalog.pg_extension SET extrelocatable = true where extname = 'vectors';
ALTER EXTENSION vectors SET SCHEMA vectors;
UPDATE pg_catalog.pg_extension SET extrelocatable = false where extname = 'vectors';
```

- When deleting pgvecto.rs, two vector columns were also deleted. So your current data may be incomplete. I'm not sure how to fix this, may need immich developer to help

**Comment by VoVAllen** [comment]
And 0.1.1 is a weird version. Immich used pgvecto.rs starting at 0.1.11. The users we encountered all upgraded from 0.1.11 or 0.1.13 to 0.2.0, and 0.1.1 is an early version that was not very stable.

**Comment by dnlldl** [comment]
I would love to use the package but most of the server world is still built around RPM packages. I'm aware a lot of people toying with Linux at home use Debian (Ubuntu and the likes) but I personally use OpenSUSE for development and otherwise SUSE and RHEL at work, both of which require RPM packages for Zypper or dnf/yum.

> To install pgvecto.rs, you don't need to build from source. We have the deb release package at https://github.com/tensorchord/pgvecto.rs/releases/tag/v0.2.0.

I will take a look at the alternatives you suggested tomorrow, thank you!

**Comment by mertalev** [comment]
> And 0.1.1 is a weird version. Immich used pgvecto.rs starting at 0.1.11. The users we encountered all upgraded from 0.1.11 or 0.1.13 to 0.2.0, and 0.1.1 is an early version that was not very stable.

I think there was a typo in the 0.1.11 release so the version shows as 0.1.1. In the upgrade we set it to 0.1.11 so it uses the right upgrade script.

**Comment by dnlldl** [comment]
> * Have you tried the command at https://docs.pgvecto.rs/admin/upgrading.html#upgrade-from-0-1-x with
> 
> ```sql
> UPDATE pg_catalog.pg_extension SET extversion = '0.1.x' where extname = 'vectors';
> UPDATE pg_catalog.pg_extension SET extrelocatable = true where extname = 'vectors';
> ALTER EXTENSION vectors SET SCHEMA vectors;
> UPDATE pg_catalog.pg_extension SET extrelocatable = false where extname = 'vectors';
> ```
> 
> * When deleting pgvecto.rs, two vector columns were also deleted. So your current data may be incomplete. I'm not sure how to fix this, may need immich developer to help

I didn't see that part of the documentation. I just tested it:

```
vm-dev:~ # su - postgres
postgres@vm-dev:~> psql
psql (15.2, server 14.10)
Type "help" for help.

postgres=# \c immich;
psql (15.2, server 14.10)
You are now connected to database "immich" as user "postgres".
immich=# UPDATE pg_catalog.pg_extension SET extversion = '0.1.1' where extname = 'vectors';
UPDATE 1
immich=# UPDATE pg_catalog.pg_extension SET extrelocatable = true where extname = 'vectors';
UPDATE 1
immich=# ALTER EXTENSION vectors SET SCHEMA vectors;
ERROR:  schema "vectors" does not exist
immich=# UPDATE pg_catalog.pg_extension SET extrelocatable = false where extname = 'vectors';
UPDATE 1
immich=# ALTER EXTENSION vectors UPDATE TO '0.2.0';
ERROR:  extension "vectors" has no update path from version "0.1.1" to version "0.2.0"

immich=# UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' where extname = 'vectors';
UPDATE 1
immich=# UPDATE pg_catalog.pg_extension SET extrelocatable = true where extname = 'vectors';
UPDATE 1
immich=# ALTER EXTENSION vectors SET SCHEMA vectors;
ERROR:  schema "vectors" does not exist
immich=# UPDATE pg_catalog.pg_extension SET extrelocatable = false where extname = 'vectors';
UPDATE 1
immich=# ALTER EXTENSION vectors UPDATE TO '0.2.0';
ERROR:  extension "vectors" has no update path from version "0.1.11" to version "0.2.0"
```

So I tried with both 0.1.1 and 0.1.11 just in case. If I remember correctly I had the `no uypdate path` error before even trying to drop the extension, that's what made me try to drop it to recreate it. Not sure where to go from there, I tried moving the `vectors.so` file as well but the database wouldn't start as it's looking for it. If there is a clean way to delete the extension completely so that I can restart from scratch, that would be great.

**Comment by dnlldl** [comment]
> > And 0.1.1 is a weird version. Immich used pgvecto.rs starting at 0.1.11. The users we encountered all upgraded from 0.1.11 or 0.1.13 to 0.2.0, and 0.1.1 is an early version that was not very stable.
> 
> I think there was a typo in the 0.1.11 release so the version shows as 0.1.1. In the upgrade we set it to 0.1.11 so it uses the right upgrade script.

Most likely this, as I definitely installed the required version to make Immich work before Immich v1.95.

**Comment by VoVAllen** [comment]
We do have upgrade script at https://github.com/tensorchord/pgvecto.rs/blob/main/sql/upgrade/vectors--0.1.11--0.2.0.sql. Can you give a try? The file seems missing with installation from cargo install .

**Comment by mertalev** [comment]
You should also run these commands to take care of the schema error:
```sql
CREATE SCHEMA IF NOT EXISTS vectors;
SET search_path TO "$user", public, vectors;
```

**Comment by dnlldl** [comment]
> We do have upgrade script at https://github.com/tensorchord/pgvecto.rs/blob/main/sql/upgrade/vectors--0.1.11--0.2.0.sql. Can you give a try? The file seems missing with installation from cargo install .

Am I supposed to put it somewhere? I tried this but that doesn't seem to do anything?

```

postgres@vm-dev:~> psql -d immich -a -f vectors--0.1.11--0.2.0.sql
\echo Use "ALTER EXTENSION vectors UPDATE TO '0.2.0'" to load this file. \quit
Use "ALTER EXTENSION vectors UPDATE TO '0.2.0'" to load this file.
```

> You should also run these commands to take care of the schema error:
> 
> ```sql
> CREATE SCHEMA IF NOT EXISTS vectors;
> ALTER DATABASE immich SET search_path TO "$user", public, vectors;
> SET search_path TO "$user", public, vectors;
> ```

Just did. Am I supposed to drop the extension after this or just try to update as normal?

```

immich=# CREATE SCHEMA IF NOT EXISTS vectors;
CREATE SCHEMA
immich=# ALTER DATABASE immich SET search_path TO "$user", public, vectors;
ALTER DATABASE
immich=# SET search_path TO "$user", public, vectors;
SET

immich=# ALTER EXTENSION vectors UPDATE TO '0.2.0';
ERROR:  extension "vectors" has no update path from version "0.1.11" to version "0.2.0"

immich=# select * from pg_available_extensions where name = 'vectors';
  name   | default_version | installed_version |                                           comment
---------+-----------------+-------------------+----------------------------------------------------------------------------------------------
 vectors | 0.2.0           | 0.1.11            | vectors: Vector database plugin for Postgres, written in Rust, specifically designed for LLM
(1 row)
```

**Comment by mertalev** [comment]
You should be able to, yes:
```sql
UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';
ALTER EXTENSION vectors SET SCHEMA vectors;
UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
ALTER EXTENSION vectors UPDATE TO '0.2.0';

SELECT pgvectors_upgrade();
```

**Comment by mertalev** [comment]
Also, if you dropped the embedding columns, you can run the commands [here](https://github.com/immich-app/immich/discussions/7310#discussioncomment-8559595) to get them back.

**Comment by dnlldl** [comment]
> You should be able to, yes:
> 
> ```sql
> UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
> UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';
> ALTER EXTENSION vectors SET SCHEMA vectors;
> UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
> ALTER EXTENSION vectors UPDATE TO '0.2.0';
> 
> SELECT pgvectors_upgrade();
> ```

```
immich=# UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
UPDATE 1
immich=# UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';
UPDATE 1
immich=# ALTER EXTENSION vectors SET SCHEMA vectors;
ALTER EXTENSION
immich=# UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
UPDATE 1
immich=# ALTER EXTENSION vectors UPDATE TO '0.2.0';
ERROR:  extension "vectors" has no update path from version "0.1.11" to version "0.2.0"
immich=# SELECT pgvectors_upgrade();
ERROR:  function pgvectors_upgrade() does not exist
LINE 1: SELECT pgvectors_upgrade();
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

I'm not even sure version 0.2.0 is correctly installed to be honest, when I downloaded and unpacked it, it showed as version `0.0.0`.

> Also, if you dropped the embedding columns, you can run the commands [here](https://github.com/immich-app/immich/discussions/7310#discussioncomment-8559595) to get them back.

```
immich=# SELECT * FROM pg_stat_all_indexes WHERE indexrelname = 'clip_index';
   relid   | indexrelid | schemaname |   relname    | indexrelname | idx_scan | idx_tup_read | idx_tup_fetch
-----------+------------+------------+--------------+--------------+----------+--------------+---------------
 240097390 |  240110157 | public     | smart_search | clip_index   |        0 |            0 |             0
(1 row)

immich=# SELECT * FROM pg_stat_all_indexes WHERE indexrelname = 'face_index';
 relid | indexrelid | schemaname |   relname   | indexrelname | idx_scan | idx_tup_read | idx_tup_fetch
-------+------------+------------+-------------+--------------+----------+--------------+---------------
 27081 |  240110158 | public     | asset_faces | face_index   |        0 |            0 |             0
(1 row)
```

**Comment by VoVAllen** [comment]
We packaged the upgrade script with the library like 
https://github.com/tensorchord/pgvecto.rs/blob/main/scripts/ci_package.sh

However, I'm not sure the installation method you used. Can you try remove the extension, and then reinstall it with the artifact from   the script above? In the current installation, the upgrade script seems missing

**Comment by dnlldl** [comment]
> We packaged the upgrade script with the library like https://github.com/tensorchord/pgvecto.rs/blob/main/scripts/ci_package.sh
> 
> However, I'm not sure the installation method you used. Can you try remove the extension, and then reinstall it with the artifact from the script above? In the current installation, the upgrade script seems missing

How do I proceed to remove it, since I'm having errors when `DROP... CASCADE` and deleting the files prevent the DB from starting? Just making sure before I commit to doing things that could be hard to reverse, I'm a system administrator but have very limited experience with PostgreSQL in general.

**Comment by VoVAllen** [comment]
Can you try the following steps?
- Drop extension CASCADE
- 'ALTER SYSTEM UNSET shared_preload_libraries'
- Restart postgres

**Comment by VoVAllen** [comment]
Before that you can try manually run https://github.com/tensorchord/pgvecto.rs/blob/main/sql/upgrade/vectors--0.1.11--0.2.0.sql which simulates the upgrade process. If it works, you don't need to reinstall the extension

**Comment by dnlldl** [comment]
> Can you try the following steps?
> 
> * Drop extension CASCADE
> * 'ALTER SYSTEM UNSET shared_preload_libraries'
> * Restart postgres

Still stuck there:

```
immich=# drop extension vectors cascade;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to column embedding of table asset_faces
drop cascades to column embedding of table smart_search
ERROR:  could not find function "vectors_amhandler_wrapper" in file "/usr/lib/postgresql14/lib64/vectors.so"
```

> Before that you can try manually run https://github.com/tensorchord/pgvecto.rs/blob/main/sql/upgrade/vectors--0.1.11--0.2.0.sql which simulates the upgrade process. If it works, you don't need to reinstall the extension

Unsure what this does.

```
postgres@vm-dev:~> psql -d immich -a -f vectors--0.1.11--0.2.0.sql
\echo Use "ALTER EXTENSION vectors UPDATE TO '0.2.0'" to load this file. \quit
Use "ALTER EXTENSION vectors UPDATE TO '0.2.0'" to load this file.
postgres@vm-dev:~> psql
psql (15.2, server 14.10)
Type "help" for help.

postgres=# \c immich;
psql (15.2, server 14.10)
You are now connected to database "immich" as user "postgres".
immich=# ALTER EXTENSION vectors UPDATE TO '0.2.0';
ERROR:  extension "vectors" has no update path from version "0.1.11" to version "0.2.0"
```

**Comment by VoVAllen** [comment]
You don't need to do `ALTER EXTENSION vectors UPDATE TO '0.2.0';`, it's the same as executing the file manually. Does your postgres work now?

**Comment by dnlldl** [comment]
> You don't need to do `ALTER EXTENSION vectors UPDATE TO '0.2.0';`, it's the same as executing the file manually. Does your postgres work with `SELECT pgvectors_upgrade();` now?

No.



> I think the first line in the script quits without running the upgrade. You can probably just remove that line and try again.

The output is a little big long and has hundreds of errors but here it is just in case:
https://pastebin.com/NXwpuSj2

I feel like at this point my installation is messy enough that having instructions on how to uninstall this thing completely and restart from scratch would be very good. Something that can get through the `DROP... CASCADE` not working above.

**Comment by mertalev** [comment]
> > You don't need to do `ALTER EXTENSION vectors UPDATE TO '0.2.0';`, it's the same as executing the file manually. Does your postgres work with `SELECT pgvectors_upgrade();` now?
> 
> No.
> 
> > I think the first line in the script quits without running the upgrade. You can probably just remove that line and try again.
> 
> The output is big and has hundreds of errors but here it is just in case: https://pastebin.com/NXwpuSj2
> 
> I feel like at this point my installation is messy enough that having instructions on how to uninstall this thing completely and restart from scratch would be very good. Something that can get through the `DROP... CASCADE` not working above.

You can try downgrading to 0.1.11 before running `DROP... CASCADE`, then upgrading.

**Comment by dnlldl** [comment]
> You can try downgrading to 0.1.11 before running `DROP... CASCADE`, then upgrading.

Apparently v0.1.11 is still working perfectly, I just changed my Immich docker repo to `ghcr.io/imagegenius/immich:v1.94.1-ig248` and it booted up just fine and is working perfectly. Both versions seem installed (???).

```
immich=# select * from pg_available_extensions where name = 'vectors';

  name   | default_version | installed_version |                                           comment                                  
---------+-----------------+-------------------+----------------------------------------------------------------------------------------------
 vectors | 0.2.0           | 0.1.11            | vectors: Vector database plugin for Postgres, written in Rust, specifically designed for LLM
(1 row)
```

However, my `/usr/lib/postgresql14/lib64/vectors.so` is the one from `/var/lib/pgsql/pgvecto.rs-0.1.11/target/release/vectors-pg14/usr/lib/postgresql14/lib64/vectors.so` (so version 0.1.11). If I put the version 0.2.0 of `vectors.so` in there, I go back to all my problems. I don't see how I can have both versions at the same time in that directory anyways?

For now I will stick with 1.94.1 until I can figure this out, at the very least things are working.

What I would need from here is pretty much live assistance to try to update pgvecto.rs to 0.2.0 correctly and figure out what doesn't work exactly. I'm having other issues when trying to properly install version 0.2.0, like mismatching versions of rust/cargo/pgrx/etc, issues with `Cargo.toml` (the new 0.2.0 version is shown as 0.0.0 by default in the zip file (???), etc). I'm clearly lacking experience here and most of the doc is very specific to users running everything in dockers. The only thing running in a docker in my case is the Immich app itself, not the database, not the extension, not Redis.

Now, is uninstall 0.2.0 as simple as deleting `/var/lib/pgsql/pgvecto.rs-0.2.0/` on my system, if I want to start the upgrade back from scratch? Should I do everything (from the rust/cargo installation to the `CREATE EXTENSION` commands) with the `postgres` user, or should some stuff be done as `root`? There are *a lot* of question marks when trying to follow the doc with an actual installation of PostgreSQL on a VM/bare-metal rather than with dockers.

**Comment by mertalev** [comment]
It sounds like you just have 0.1.11 installed right now. I think your basic goal should be to successfully drop the extension just to have a clean slate.

The issue in the OP about dropping it is that you basically can't interact with the extension on 0.2.0 until you do the upgrade. This includes dropping it. You should be able to drop it now if you have 0.1.11 installed.

After that, delete the pg_vectors folder and you'll have a clean environment to install 0.2.0.

**Comment by dnlldl** [comment]
> It sounds like you just have 0.1.11 installed right now. I think your basic goal should be to successfully drop the extension just to have a clean slate.
> 
> The issue in the OP about dropping it is that you basically can't interact with the extension on 0.2.0 until you do the upgrade. This includes dropping it. You should be able to drop it now if you have 0.1.11 installed.
> 
> After that, delete the pg_vectors folder and you'll have a clean environment to install 0.2.0.

Well that was... easy?
```
immich=# drop extension vectors cascade;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to column embedding of table asset_faces
drop cascades to column embedding of table smart_search
DROP EXTENSION
```

Then:
```
vm-dev:~ # systemctl stop postgresql
vm-dev:~ # cp /var/lib/pgsql/pgvecto.rs-0.2.0/target/release/vectors-pg14/usr/lib/postgresql14/lib64/vectors.so /usr/lib/postgresql14/lib64/vectors.so
vm-dev:~ # systemctl start postgresql
```

Then:
```
immich=# create extension vectors;
CREATE EXTENSION
```

Updated immich to v1.95.1 and it starts! now, saw errors when searching, such as:
```
QueryFailedError: column search.embedding does not exist
```

So I did this:
```
immich=# SELECT * FROM pg_stat_all_indexes WHERE indexrelname = 'clip_index';
 relid | indexrelid | schemaname | relname | indexrelname | idx_scan | idx_tup_read | idx_tup_fetch
-------+------------+------------+---------+--------------+----------+--------------+---------------
(0 rows)

immich=# SELECT * FROM pg_stat_all_indexes WHERE indexrelname = 'face_index';
 relid | indexrelid | schemaname | relname | indexrelname | idx_scan | idx_tup_read | idx_tup_fetch
-------+------------+------------+---------+--------------+----------+--------------+---------------
(0 rows)

immich=# DELETE FROM smart_search WHERE true;
DELETE 15545
immich=# DELETE FROM asset_faces WHERE true;
DELETE 9489
immich=# DELETE FROM person WHERE true;
DELETE 2830
immich=# ALTER TABLE smart_search
immich-# ADD COLUMN IF NOT EXISTS embedding vector(512);
ALTER TABLE
immich=# ALTER TABLE asset_faces
immich-# ADD COLUMN IF NOT EXISTS embedding vector(512);
ALTER TABLE
immich=# SET vectors.pgvector_compatibility=on;
SET
immich=# CREATE INDEX IF NOT EXISTS clip_index ON smart_search
immich-#     USING hnsw (embedding vector_cosine_ops)
immich-#     WITH (ef_construction = 300, m = 16);
ERROR:  permission denied for schema vectors
immich=# alter role immich with superuser;
ALTER ROLE
immich=# CREATE INDEX IF NOT EXISTS clip_index ON smart_search
    USING hnsw (embedding vector_cosine_ops)
    WITH (ef_construction = 300, m = 16);
CREATE INDEX
immich=# CREATE INDEX IF NOT EXISTS face_index ON asset_faces
immich-#     USING hnsw (embedding vector_cosine_ops)
immich-#     WITH (ef_construction = 300, m = 16);
CREATE INDEX
```
Then ran face detection/recognition and smart search smart learning

Everything seems to be working as normal now. So long as you don't mind losing facial recognition data (not an issue for me).

Now... I'm not entirely sure where I screwed up initially. I think dropping the extension BEFORE updating is all in all safer? And it seems like there is no need for any upgrade script? (I did try `SELECT pgvectors_upgrade();` again hoping it would fix the smart search issues, it did its thing but didn't help for smart search.)

Also note that the `superuser` permission is required, otherwise https://github.com/immich-app/immich/issues/7289.

**Comment by VoVAllen** [comment]
After discussing with team member, the error may occur when the extension is not properly upgraded. To solve this, you need to copy the file `https://github.com/tensorchord/pgvecto.rs/blob/main/sql/upgrade/vectors--0.1.11--0.2.0.sql` to the postgres extension folder.

The postgres extension folder usually locate at `/usr/share/postgresql/14/extension`. If it's not the case on your system, you can use `SHOW config_file;` to find the config path. Open this configuration file, look for the value of data_directory, which will tell you where the database's data directory is located. The extension directory is usually located at a relative path like ../share/extension within the data directory or a similar location, depending on your PostgreSQL installation.

After copy the file into the postgres extension folder, you should be able to run `ALTER EXTENSION vectors UPDATE TO '0.2.0';`

**Comment by dnlldl** [comment]
> After discussing with team member, the error may occur when the extension is not properly upgraded. To solve this, you need to copy the file `https://github.com/tensorchord/pgvecto.rs/blob/main/sql/upgrade/vectors--0.1.11--0.2.0.sql` to the postgres extension folder.
> 
> The postgres extension folder usually locate at `/usr/share/postgresql/14/extension`. If it's not the case on your system, you can use `SHOW config_file;` to find the config path. Open this configuration file, look for the value of data_directory, which will tell you where the database's data directory is located. The extension directory is usually located at a relative path like ../share/extension within the data directory or a similar location, depending on your PostgreSQL installation.
> 
> After copy the file into the postgres extension folder, you should be able to run `ALTER EXTENSION vectors UPDATE TO '0.2.0';`

I think you replied before you saw my comment, right? Because right now I'm on 0.2.0 so it wouldn't make sense (and it would not work) to `ALTER EXTENSION vectors UPDATE TO '0.2.0';`.

**Comment by dnlldl** [comment]
If I may add, I just noticed when you install the extension (with `cargo pgrx install -c /usr/lib/postgresql14/bin/pg_config --sudo --release` in my case), it does copy the `vectors.so` and `vectors.control` file correctly. The `cp` I did was just because I was basically switching back and forth between the 2 versions (which should not be necessary and I definitely do not recommend).

All in all, I think next time rather than trying to upgrade, I will just remove the extension, stop the database, install the new version, boot the database and enable the new extension. Seems like it's just much easier this way... but that may require some additional testing, and I guess the upgrade scripts you guys worked take care of some other stuff that could have prevented me from losing my smart search and facial recog stuff.

**Comment by VoVAllen** [comment]
> > After discussing with team member, the error may occur when the extension is not properly upgraded. To solve this, you need to copy the file `https://github.com/tensorchord/pgvecto.rs/blob/main/sql/upgrade/vectors--0.1.11--0.2.0.sql` to the postgres extension folder.
> > The postgres extension folder usually locate at `/usr/share/postgresql/14/extension`. If it's not the case on your system, you can use `SHOW config_file;` to find the config path. Open this configuration file, look for the value of data_directory, which will tell you where the database's data directory is located. The extension directory is usually located at a relative path like ../share/extension within the data directory or a similar location, depending on your PostgreSQL installation.
> > After copy the file into the postgres extension folder, you should be able to run `ALTER EXTENSION vectors UPDATE TO '0.2.0';`
> 
> I think you replied before you saw my comment, right? Because right now I'm on 0.2.0 so it wouldn't make sense (and it would not work) to `ALTER EXTENSION vectors UPDATE TO '0.2.0';`.

Yes. Glad to hear you've found the solution!

**Comment by VoVAllen** [comment]
> If I may add, I just noticed when you install the extension (with `cargo pgrx install -c /usr/lib/postgresql14/bin/pg_config --sudo --release` in my case), it does copy the `vectors.so` and `vectors.control` file correctly. The `cp` I did was just because I was basically switching back and forth between the 2 versions (which should not be necessary and I definitely do not recommend).
> 
> All in all, I think next time rather than trying to upgrade, I will just remove the extension, stop the database, install the new version, boot the database and enable the new extension. Seems like it's just much easier this way... but that may require some additional testing, and I guess the upgrade scripts you guys worked on take care of some other stuff that could have prevented me from losing my smart search and facial recog stuff.

We didn't realize the upgrade SQL is missing when install from the source code. In the next release, we will provide a tar package including binary libraries and all upgrade and related files, as well as a simple guide to help non-Debian users install from binaries. Thanks for your report and patience with us!

**Comment by dnlldl** [comment]
> We didn't realize the upgrade SQL is missing when install from the source code. In the next release, we will provide a tar package including binary libraries and all upgrade and related files, as well as a simple guide to help non-Debian users install from binaries. Thanks for your report and patience with us!

Thanks for your patience with me, glad to help and learn at the same time.

Also, I know nothing about pgrx extensions in PostgreSQL so I may be over my head here, but I had to edit the supplied `Cargo.toml` file from the source code to change the extension version from 0.0.0 to 0.2.0 (otherwise it would just not work) and I had to change the pgrx prerequisite as well from 0.11.2 to the version installed on my system, 0.11.3. Basically, those 2 lines had to be edited from the source code for me to work:
```
[dependencies]
pgrx = { version = "0.11.2", default-features = false, features = [] }

[workspace.package]
version = "0.0.0"
```
That is from what is supplied in https://github.com/tensorchord/pgvecto.rs/archive/refs/tags/v0.2.0.zip.

I also had an error (which I ended up just ignoring) when installing saying the code was built with rust 1.77-nightly while I have rust 1.76 stable (installed with https://sh.rustup.rs). The error basically says the installation of the extension failed... but everything worked.
